File: algorithms/160 - Intersection of Two Linked Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    iter(headA);
    return iter(headB);
};

const iter = (head) => {
    while (head) {
        if (head.visited) return head;
        head.visited = true;
        head = head.next;
    }
    return null;
}

File: algorithms/ruby/160 - Intersection of Two Linked Lists.rb
require 'set'
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} headA
# @param {ListNode} headB
# @return {ListNode}
def getIntersectionNode(headA, headB)
  set = Set.new
  while headA do
    set << headA
    headA = headA.next
  end
  while headB do
    return headB if set === headB
    headB = headB.next
  end
end
